src/mlpack/tests/main_tests/fastmks_test.cpp

```diff
@@ -0,0 +1,658 @@
+{
+  // 100 points in 3 dimensions.
+  arma::mat referenceData(3, 100, arma::fill::randu);
+
+  // Random input, some k <= number of reference points.
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("k", (int) 10);
+
+  mlpackMain();
+
+  // Check the indices matrix has 10 points for each input point.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::Mat<size_t>>
+      ("indices").n_rows, 10);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::Mat<size_t>>
+      ("indices").n_cols, 100);
+
+  // Check the kernel matrix has 10 points for each input point.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("kernels").n_rows, 10);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("kernels").n_cols, 100);
+}
+
+/**
+ * Ensure that saved model can be used again.
+ */
+BOOST_AUTO_TEST_CASE(FastMKSModelReuseTest)
+{
+  // 100 points in 3 dimensions.
+  arma::mat referenceData(3, 100, arma::fill::randu);
+  // 90 points in 3 dimensions.
+  arma::mat queryData(3, 90, arma::fill::randu);
+
+  // Random input, some k <= number of reference points.
+  SetInputParam("reference", referenceData);
+  SetInputParam("query", queryData);
+
+  mlpackMain();
+
+  arma::Mat<size_t> indices;
+  arma::mat kernel;
+  FastMKSModel* output_model;
+  indices = std::move(CLI::GetParam<arma::Mat<size_t>>("indices"));
+  kernel = std::move(CLI::GetParam<arma::mat>("kernels"));
+  output_model = std::move(CLI::GetParam<FastMKSModel*>("output_model"));
+
+  // Reset passed parameters.
+  CLI::GetSingleton().Parameters()["reference"].wasPassed = false;
+  CLI::GetSingleton().Parameters()["query"].wasPassed = false;
+
+  // Input saved model, pass the same query and keep k unchanged.
+  SetInputParam("input_model", output_model);
+  SetInputParam("query", queryData);
+
+  mlpackMain();
+
+  // Check that initial output matrices and the output matrices using
+  // saved model are equal.
+  CheckMatrices(indices, CLI::GetParam<arma::Mat<size_t>>("indices"));
+  CheckMatrices(kernel, CLI::GetParam<arma::mat>("kernels"));
+}
+
+/*
+ * Ensure that reference dataset gives the same result when passed as
+ * a query dataset
+ */
+BOOST_AUTO_TEST_CASE(FastMKSQueryRefTest)
+{
+  // 100 points in 3 dimensions.
+  arma::mat referenceData(3, 100, arma::fill::randu);
+  // 100 points in 3 dimensions.
+  arma::mat queryData(3, 100, arma::fill::randu);
+
+  // Random input, some k <= number of reference points.
+  SetInputParam("reference", referenceData);
+  SetInputParam("query", queryData);
+  SetInputParam("k", (int) 10);
+
+  mlpackMain();
+
+  arma::Mat<size_t> indices;
+  arma::mat kernel;
+  indices = std::move(CLI::GetParam<arma::Mat<size_t>>("indices"));
+  kernel = std::move(CLI::GetParam<arma::mat>("kernels"));
+
+  bindings::tests::CleanMemory();
+
+  CLI::GetSingleton().Parameters()["reference"].wasPassed = false;
+  CLI::GetSingleton().Parameters()["query"].wasPassed = false;
+
+
+  SetInputParam("reference", referenceData);
+  SetInputParam("query", queryData);
```

Comment:
I'm a bit confused here, you've passed the same dataset twice.  I thought the idea was to check that we get the same results when we leave `query` unset, so the query set is the reference set.

Also, technically, the results will not be the same when that is done: if we specify `'reference'` as `referenceSet only, then each point cannot appear in its own results.  But if we specify `'reference'` as `referenceSet` and also `'query'` as `referenceSet`, then, e.g., point 0 can appear (and is likely to) in the top-k max-kernel results for point 0.  However, you could run the test, and then in the comparison of any results, account for this by skipping over any cases where the point appears in its own results.  Let me know if I can clarify that point.

Reply:
Please let me know that if I have done some correct or I making some big blunder here. :)